### PR TITLE
new function for getting the datatypes of the column

### PR DESCRIPTION
**Pull request by susgo**
Hi,

if this feature is not needed for anyone else please feel free to close this pull request.

In some cases our customers need the information about the datatypes given in their csv file and they can change it, if it doesn't fit (column-wise). So they need it for some comparative analysis before data mapping.

If you like this feature I will improve the code snippets. Therefore feel free to review the existing code. I would be grateful to you for any feedback, suggestions, notes, questions but also constructive criticism as I allways want to learn. :-) 

Thanks!

### CODE REVIEW

**Comment by Fonata** [comment]
Personally, I do think data type information is a valid and useful extension for this library.

Here's what's still needed in my view:
- since the class is already rather long, please extract your new code into its own trait
- update the "worthless" test with the new fields
- replace "Commonly" with "Common" or "Frequent" in the function name, as the phrase doesn't look like an adverb to me.
- PHPDoc types: please remove the brackets [] around the types
- code formatting: make sure there is a space after each comma. PhpStorm will help you with that.

**Comment by susgo** [comment]
Thanks for reviewing my code. Changes done. See also #115 for some future improvements.

File: extensions/Datatype.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+     * Check for most commonly data type for one column.
+     *
+     * @access private
+     *
+     * @param  [array] $datatypes
+     *
+     * @return [string|false]
+     */
+    private function getMostFrequentDataypeForColumn($datatypes) {
+        if (isset($datatypes['unknown'])) {
```
Comment (Fonata) [inline extensions/Datatype.php]: if I am not mistaken this `if` is not necessary, you can just use `unset` (even if the array key doesn't exist)

File: extensions/Datatype.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
+<?php
+
+trait Datatype {
```
Comment (Fonata) [inline extensions/Datatype.php]: Please add the suffix Trait here (and in the filename)

File: extensions/Datatype.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Check data type for one column.
+     * Check for most commonly data type for one column.
+     *
+     * @access private
+     *
+     * @param  [array] $datatypes
+     *
+     * @return [string|false]
```
Comment (Fonata) [inline extensions/Datatype.php]: PHPDoc: I like that you've added documentation blocks. Please remove the [] to keep the style compatible to the rest of this project's code.

File: extensions/Datatype.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        foreach ($this->titles as $cName) {
+            $column = array_column($this->data, $cName);
+            $cDatatypes = array_map([$this, 'getDatatypeFromString'], $column);
+
+            $result[$cName] = $this->getMostFrequentDataypeForColumn($cDatatypes);
+        }
+
+        $this->data_types = $result;
+
+        return !empty($this->data_types) ? $this->data_types : FALSE;
```
Comment (Fonata) [inline extensions/Datatype.php]: Code style: the rest of the code base uses lower-case constants. You can use [this PhpStorm code style archive](https://github.com/parsecsv/parsecsv-for-php/issues/101#issuecomment-355733791) to speed things up. Thanks!

File: extensions/Datatype.php (hunk 1 above)
Comment (susgo) [reply extensions/Datatype.php]: Really not necessary. Changed it.

File: extensions/Datatype.php (hunk 2 above)
Comment (susgo) [reply extensions/Datatype.php]: I'm not a friend of such suffix' but have changed it.

File: extensions/Datatype.php (hunk 3 above)
Comment (susgo) [reply extensions/Datatype.php]: done.

File: extensions/Datatype.php (hunk 4 above)
Comment (susgo) [reply extensions/Datatype.php]: done.... and I miss the GitLab `Resolve`-Button here in GitHub for this situation 😟

**Comment by Fonata** [comment]
Wow, you are quick! The Trait naming system is standardized in PSR: https://www.php-fig.org/bylaws/psr-naming-conventions/

**Comment by Fonata** [comment]
@jimeh @williamknauss I am happy with the changes proposed by **susgo**.

As usually, I'll merge this pull request in 7 days unless I hear from you guys until then. Feel free to share your thoughts if you find the time. If not: no pressure.

**Comment by susgo** [comment]
One thing is also missing or bad done in this pull request. Before merging this I would also introduce some constants for the datatypes. Therefore cleanest way is to introduce new enum class holding this special constants - in my opinion. Are there any naming rules for such classes, e.g. prefix or suffix `Enum`?

**Comment by Fonata** [comment]
Good morning @susgo! No, I am not aware of a specialized type for this. 

My preferred way is to group constants in a class. The suffix ...Enum would be understood by most programmers. Please extend `SplEnum`.

**Comment by susgo** [comment]
new enum class follows in new pull request as I want to introduce namespaces first. First steps done on our fork but not ready yet because I have had problems running the tests

**Comment by Fonata** [comment]
Regarding tests: locally or the Travis tests by GitHub? I’d love to help.

**Comment by susgo** [comment]
Locally, I have added the exception to the pull request on our itexia fork because I do not have more time for this yet. I think, I have done something easy wrong.

**Comment by Fonata** [comment]
I use a phpunit.phar locally. I‘ll send you the exact version and command line that PhpStorm uses. I’ll post it with the picture here until 9pm tonight

**Comment by susgo** [comment]
More clearly: the tests were running perfectly before the namespace changes I have done on itexia namespace branch.

**Comment by susgo** [comment]
In regards to `SplEnum` I would like to have some feedback from the community. I never used it because I use different package for our framework. I don't like to use it as it is not build-in into PHP and requires pecl extension installed. Could not be included via composer and is not pre installed. Do I miss something? Why is it better?

**Comment by Fonata** [comment]
@susgo Please ignore my immature suggestions on `SplEnum`. Wow, PECL is a no-go for me! I didn't realize that it is not there by default.

**Comment by Fonata** [comment]
A simple class is enough then.

**Comment by susgo** [comment]
This is up to date now. But tests not working for me locally

File: composer.json [hunk 5]
```diff
@@ -12,10 +12,12 @@
...
     ],
-    "autoload": {
-        "classmap": [
-            "."
-        ]
+    "autoload":{
+        "psr-4":{
+            "ParseCsv\\": "src",
+            "ParseCsv\\extensions\\": "src\\extensions",
+            "ParseCsv\\tests\\": "tests"
```
Comment (susgo) [inline composer.json]: should be moved to section "autoload-dev"

**Comment by susgo** [comment]
Change is ready except for one test: `ParseCsv\tests\methods\ParseTest::testGetColumnDatatypes`  Thats why I have written some additional tests and I have tested it manually on our project. I cannot reproduce this. Maybe it is a little thing in the test csv I miss at the moment - sorry for that 😟 

The pull request contains now:
* the new functionallity introduced by a trait and an initial enum class for constant handling.
* introduces namespaces
* new tests

File: composer.json (hunk 5 above)
Comment (Fonata) [reply composer.json]: Yeah, you're right. I saw this now in a lot of open source libraries. Fixed in https://github.com/itexia/parsecsv-for-php/pull/6.

File: src/enums/DatatypeEnum.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * Class DatatypeEnum
+ *
+ * @package ParseCsv\enums
+ *
+ * todo: needs a basic parent enum class for error handling.
+ */
+class DatatypeEnum
+{
```
Comment (Fonata) [inline src/enums/DatatypeEnum.php]: The curly bracket should go to the line above to match the style of the rest of the project. I fixed it in https://github.com/itexia/parsecsv-for-php/pull/6.

**Comment by Fonata** [comment]
Yeah, finally [all tests are green](https://travis-ci.org/parsecsv/parsecsv-for-php/builds/344032671)!

**Pull request merged**